rfcs/InputUnion.md

```diff
@@ -18,8 +18,119 @@ To help bring this idea to reality, you can contribute through two channels:
+
+Let's imagine an animal shelter for our example. When querying for a list of the animals, it's easy to see how abstract types are useful - we can get data specific to the type of the animal easily.
+
+```graphql
+{
+  animalShelter(location: "Portland, OR") {
+    animals {
+      __typename
+      name
+      age
+      ... on Cat { livesLeft }
+      ... on Dog { breed }
+      ... on Snake { venom }
+    }
+  }
+}
+```
+
+However, when we want to submit data, we can't use an `interface` or `union`, so we must model around that.
+
+One technique commonly used to is a **Tagged union** pattern. This essentially boils down to a "wrapper" input that isolates each type into it's own field. The field name takes on the convention of representing the type.
+
+```graphql
+mutation {
+  logAnimalDropOff(
+    location: "Portland, OR"
+    animals: [
+      {cat: {name: "Buster", age: 3, livesLeft: 7}}
+    ]
+  )
+}
+```
+
+Unfortunately, this opens up a set of problems, since the Tagged union input type actually contains many fields, any of which could be submitted.
+
+```graphql
+input AnimalDropOff {
+  cat: Cat
+  dog: Dog
+  snake: Snake
+}
+```
+
+This leads to non-sensical mutations that are totally valid, like an animal that is both a `Cat` and a `Dog`
+
+```graphql
+mutation {
+  logAnimalDropOff(
+    location: "Portland, OR"
+    animals: [
+      {
+        cat: {name: "Buster", age: 3, livesLeft: 7},
+        dog: {name: "Ripple", age: 2, breed: WHIPPET}
+      }
+    ]
+  )
+}
+```
+
+In addition, relying on this layer of abstraction means that this domain is modeled differently across input & output. This puts a large burden on the developer interacting with the API, as they have to code against this.
+
+```json
+// JSON structure returned from a query
+{
+  "animals": [
+    {"__typename": "Cat", "name": "Ruby", "age": 2, "livesLeft": 9}
+    {"__typename": "Snake", "name": "Monty", "age": 13, "venom": "POISON"}
+  ]
+}
+```
+
+```json
+// JSON structure submitted to a mutation
+{
+  "animals": [
+    {"cat": {"name": "Ruby", "age": 2, "livesLeft": 9}},
+    {"snake": {"name": "Monty", "age": 13, "venom": "POISON"}}
+  ]
+}
+```
+
+Another common approach is to provide a unique mutation for every type. So you'd have a `logCatDropOff` and `logDogDropOff` and `logSnakeDropOff`. That removes the potential for modeling non-sensical situations, but it explodes the number of mutations in an API.
+
+These workarounds only get worse at scale. Real world APIs can have dozens if not hundreds of possible types.
+
+The idea of the **Input Union** is to bring an abstract type to inputs. This would enable us to model our domain as elegantly in input as we can in output.
```

Comment:
I think the "would [...] as elegantly" claim might be a little strong; IMO none of the proposed solutions live up to this. I've softened it marginally:

```suggestion
The idea of the **Input Union** is to bring an abstract type to inputs. This would enable us to model situations where an input may be of different types in a type-safe and elegant manner, like we can with outputs.
```